src/storage/storage-async.ts

```diff
@@ -189,141 +200,161 @@ export class StorageAsync implements IStorageAsync {
+                    return {
+                        kind: 'nothing_happened',
+                        reason: 'already_had_it',
+                        doc: docToIngest,
+                        maxLocalIndex: this.storageDriver.getMaxLocalIndex(),
+                    };
+                }
             }
 
             // save it
             loggerIngest.debug('  > upserting into storageDriver...');
-            let docResult = await this.storageDriver.upsert(docToIngest);
+            let docAsWritten = await this.storageDriver.upsert(docToIngest);  // TODO: pass existingDocsSamePath to save another lookup
             loggerIngest.debug('  > ...done upserting into storageDriver');
-            loggerIngest.debug(' >> ingest: end of protected region');
-
-            if (!docResult) { return { ingestResult: IngestResult.WriteError, docIngested: null}; }
-            return isLatest
-                ? { ingestResult: IngestResult.AcceptedAndLatest, docIngested: docResult }
-                : { ingestResult: IngestResult.AcceptedButNotLatest, docIngested: docResult }
+            loggerIngest.debug('  > ...getting storageDriver maxLocalIndex...');
+            let maxLocalIndex = this.storageDriver.getMaxLocalIndex();
+
+            loggerIngest.debug(' >> ingest: end of protected region, returning a WriteEvent from the lock');
+            return {
+                kind: 'success',
+                maxLocalIndex,
+                doc: docAsWritten,  // with updated extra properties like _localIndex
+                docIsLatest: isLatest,
+                prevDocFromSameAuthor: prevSameAuthor,
+                prevLatestDoc: prevLatest,
+            };
         };
 
         loggerIngest.debug(' >> ingest: running protected region...');
-        // bypass the lock if we are already in a lock (because we're being called from set())
-        let result: IngestResultAndDoc = await this.storageDriver.lock.run(protectedCode, { bypass: _bypassLock });
-        let { ingestResult, docIngested } = result;
-        loggerIngest.debug(' >> ingest: ...done running protected region', ingestResult);
+        let ingestEvent: IngestEvent = await this._ingestLock.run(writeToDriverWithLock);
+        loggerIngest.debug(' >> ingest: ...done running protected region');
 
-        if (this._isClosed) { throw new StorageIsClosedError(); }
+        loggerIngest.debug('...send ingest event after releasing the lock');
+        await this.bus.sendAndWait('ingest', ingestEvent);
 
-        // only send events if we successfully ingested a doc
-        if (docIngested !== null) {
-            loggerIngest.debug('  - ingest: send ingest event blockingly...');
-            await this.bus.sendAndWait(('ingest|' + docIngested.path) as 'ingest', docIngested);
-            loggerIngest.debug('  - ingest: ...done sending ingest event');
-        }
-
-        return result;
+        return ingestEvent;
```

Comment:
Now we send `IngestEvents` to the bus even if it was unsuccessful or nothing happened (e.g. the document was obsolete, we already had it, validation error, etc)